### PR TITLE
Fix query string binding

### PR DESCRIPTION
**Pull request by mkurz**
As [promised](https://github.com/playframework/playframework/pull/8844#issuecomment-442668402).

Just have a look at the tests. I am pretty sure that's how we want the various `QueryStringBindable`s to behave and before they were broken (yes, just for the empty string `?x=` cases, but still).

(You can ignore the first two commits, just cleanup)

### CODE REVIEW

File: framework/src/sbt-plugin/src/sbt-test/routes-compiler-plugin/routes-compilation/tests/RouterSpec.scala [hunk 1]
```diff
@@ -92,7 +92,7 @@ object RouterSpec extends PlaySpecification {
                                     withDefault: Boolean = false) = {
     lazy val resolvedPath = s"/${path}${if(withDefault) "-d" else ""}"
     s"bind ${paramType} parameter${if(withDefault) " with default value" else ""} from the query string" in {
-      "succesfully" in new WithApplication() {
+      "successfully" in new WithApplication() {
```
Comment (marcospereira) [inline framework/src/sbt-plugin/src/sbt-test/routes-compiler-plugin/routes-compilation/tests/RouterSpec.scala]: 👍

File: framework/src/play/src/main/scala/play/core/routing/GeneratedRouter.scala [hunk 2]
```diff
@@ -66,9 +70,16 @@ case class RouteParams(path: Map[String, Either[Throwable, String]], queryString
   }
 
   def fromQuery[T](key: String, default: Option[T] = None)(implicit binder: QueryStringBindable[T]): Param[T] = {
-    Param(key, binder.bind(key, queryString).getOrElse {
-      default.map(d => Right(d)).getOrElse(Left("Missing parameter: " + key))
-    })
+    val bindResult = binder.bind(key, queryString)
```
Comment (marcospereira) [inline framework/src/play/src/main/scala/play/core/routing/GeneratedRouter.scala]: See if declaring a `val flatBindResult = bindResult.flatMap(_.toOption)` can make your life easier at the `if` below.

Comment (mkurz) [reply framework/src/play/src/main/scala/play/core/routing/GeneratedRouter.scala]: Hmm.. I can't see how that would help... Or maybe I am missing something?

**Comment by mkurz** [comment]
Thanks @marcospereira! 
1. No, no need for the commented tests that were removed. I kind of generated the tests (the routes, method signatures), and ended up with test cases which don't make much sense (like a Java `Optional` wrapping a Scala `Char`, etc.).
2. Migration notes added.

Please have a look again, thanks!

File: documentation/manual/releases/release27/migration27/Migration27.md [hunk 3]
```diff
@@ -518,3 +518,40 @@ libraryDependencies += "org.apache.commons" % "commons-lang3" % "3.8.1"
 ### Evolutions comment syntax
 
 Play Evolutions now correctly supports SQL92 comment syntax. This means you can write evolutions using `--` at the beginning of a line instead of `#` wherever you choose. Newly generated evolutions using the Evolutions API will now also use SQL92-style comment syntax in all areas. Documentation has also been updated accordingly to prefer the SQL92 style, though the older comment style is still fully supported.
+
+### Query string parameter binding behaviour changed
+
+#### When a param's value is empty (e.g `?myparam=`)
+
+Routes that define query string parameters of the types
```
[the remark this answers is not in the document]
Comment (mkurz) [reply documentation/manual/releases/release27/migration27/Migration27.md]: You probably mean `the following types` :wink:

**Comment by mkurz** [comment]
@marcospereira Updated, thanks!

**Comment by mkurz** [comment]
@marcospereira Updated, thanks!

**Comment by marcospereira** [comment]
Backport to 2.7.x: ac385695fba9d4be28a1a7b272f5b0830a383652

**Comment by mbryzek** [comment]
Thanks! This is a nice bug fix and will remove workarounds we have - much appreciated!

**Pull request merged**